docs/internal/abi.md

```diff
@@ -0,0 +1,409 @@
+
+Example:
+```ts
+// NaN-boxed types
+Array[U16 | I8] -> [ i64, i64, i64, ... ]
+Array[U32 | F64] -> [ i64, i64, i64, ... ]
+Array[A | F16] -> [ i64, i64, i64, ... ]
+
+// wide-packed types
+Array[A | U64] -> [ { i8, i64 }, { i8, i64 }, ... ]
+Array[F64 | U128] -> [ { i8, i128 }, { i8, i128 }, ... ]
+```
+
+NaN-boxed types have the problem that it increases the size of the object at least to 64-bits (128-bit on CHERI), even when neither of the types are 64-bit wide.
+This may not be a problem for individual objects, but an array with multiple objects will multiply the problem.
+
+Example:
+```ts
+// Single-types
+Array[I8] -> [ i8, i8, i8, ... ]
+Array[U8] -> [ i8, i8, i8, ... ]
+
+// NaN-boxed types
+Array[U8 | I8] -> [ i64, i64, i64, ... ]
+```
+
+That, however, is a small problem when compared to the wide packing.
+Not only the wide-packed array increases the payload, but it also introduces the descriptor which is not the same size as the payload.
+This creates misalignment issues (on platforms where misaligned reads are penalised or forbidden, this is really bad).
+
+An optimisation under discussion is to pack the descriptors into a payload-sized element.
+With a 64-bit payload, the first 8 elements would follow the first 64-bit bucket with their 8 descriptors packed.
+The following 8 elements would need a new descriptor bucket, placed after the 8th element, and so on.
+
+Example:
+```ts
+Array[F64 | U128] -> [ { i8, i64 }, { i8, i64 }, ... ]
+
+// to
+
+type Desc = { i8, i8, i8, i8, i8, i8, i8, i8 };
+Array[F64 | U128] -> [ Desc, i64, i64, i64, i64, i64, i64, i64, i64, Desc, i64, ...  ]
+```
+
+To access the 7-th element, the unpacking does:
+```ts
+// Takes the 7th descriptor (in pseudo tuple notation starting at _0)
+descrip = Array[0]._6;
+
+// Takes the 7th element (at 8th position starting from 0)
+payload = Array[6+1];
+```
+
+These problems only happens in arrays of unions that need to be wide-packed.
+Users should prefer arrays of concrete types as much as possible for code that needs to be fast (inner loops, etc).
+
+## Calling Convention
+
+The Verona compiler must follow the target's calling convention, which could be a particular hardware, some sandbox technology, or calling external (Verona) code via shared libraries.
+
+However, the compiler is allowed to change the encoding and order of the arguments and even remove completely some arguments that are shown not to be used, in order to improve performance, while still fitting to the calling convention used.
+
+Luckily, by using LLVM, we don't need to worry about any calling convention features that "just work" and only need to look at issues if they prove to be a case common enough to show up in profiles.
+
+This section discusses some of those opportunities.
+
+### Concrete Types
+
+FIXME: The wording below is confusing, please suggest a better one.
+
+Concrete types, including pointers, are always passed by value to functions.
```

Comment:
I think this paragraph is correct, but poorly worded. I may not be using the correct Verona terminology here.